### PR TITLE
Remove AspNetCoreModule virtual projects

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

**Comment by jkotalik** [comment]
How does this affect site extensions?

**Comment by pakrym** [comment]
Site extensions don't use `AspNetCoreModule` project reference. It uses direct vcxproj reference instead. https://github.com/aspnet/AspNetCore/blob/master/src/SiteExtensions/Runtime/Microsoft.AspNetCore.Runtime.SiteExtension.pkgproj#L30

File: src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj [hunk 1]
```diff
@@ -10,8 +10,16 @@
...
   <ItemGroup>
     <PackageReference Remove="Internal.AspNetCore.Sdk" />
-    <PackageReference Include="Microsoft.AspNetCore.AspNetCoreModule" Version="$(MicrosoftAspNetCoreAspNetCoreModulePackageVersion)" />
-    <PackageReference Include="Microsoft.AspNetCore.AspNetCoreModuleV2" Version="$(MicrosoftAspNetCoreAspNetCoreModuleV2PackageVersion)" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV1\AspNetCore\AspNetCore.vcxproj" Platform="x64" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV1\AspNetCore\AspNetCore.vcxproj" Platform="x86" />
+
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj"  Platform="x64" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj" HandlerPath="2.0.0"  Platform="x64" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj"  Platform="x64" />
```
Comment (jkotalik) [inline src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj]: You shouldn't need Inproc.

Comment (pakrym) [inline src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj]:
```suggestion
```

File: src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj [hunk 2]
```diff
@@ -10,8 +10,16 @@
...
+
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj" Platform="x86" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj" HandlerPath="2.0.0" Platform="x86" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="x86" />
```
Comment (pakrym) [inline src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj]:
```suggestion
```

File: src/Servers/IIS/ResolveIisReferences.targets [hunk 3]
```diff
@@ -8,23 +8,6 @@ with the right MSBuild incantations to get output copied to the right place.
   <Choose>
     <!-- IIS native projects only build on Windows with MSBuild.exe -->
     <When Condition="'$(OS)' == 'Windows_NT' AND '$(MSBuildRuntimeType)' != 'Core' ">
-      <ItemGroup Condition=" '@(Reference->AnyHaveMetadataValue('Identity', 'AspNetCoreModule'))' == 'true' ">
```
Comment (jkotalik) [inline src/Servers/IIS/ResolveIisReferences.targets]: cc @natemcmaster with regards to making this work.

File: src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj (hunk 2 above)
Comment (jkotalik) [reply src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj]: u lazy 😆

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (pakrym) [reply src/Servers/IIS/ResolveIisReferences.targets]: Making what work?

**Review by jkotalik — APPROVED** [review]
LGTM pending Reference vs NativeProjectReference.

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (jkotalik) [reply src/Servers/IIS/ResolveIisReferences.targets]: I have never seen `<Reference Include="AspNetCoreModule">` work correctly (it fails on restore). I'm fine with removing it, but it would be nice to have a shortcut for referencing native projects.

Comment (pakrym) [reply src/Servers/IIS/ResolveIisReferences.targets]: Problem with the shortcut is that it's almost never correct. You never need both handler and shim references from one project.

**Comment by pakrym** [comment]
Installer part needs a bit more work to remove ANCMPackageResolver step.

**Comment by jkotalik** [comment]
>Installer part needs a bit more work to remove ANCMPackageResolver step.

I believe I'm assigned to cleanup the ANCM installers eventually 😄

**Comment by pakrym** [comment]
Want to take this over?

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (jkotalik) [reply src/Servers/IIS/ResolveIisReferences.targets]: I'm thinking of the following:
```msbuild
<ItemGroup Condition=" '@(Reference->AnyHaveMetadataValue('Identity', 'AspNetCoreModule'))' == 'true' ">
        <Reference Remove="AspNetCoreModuleV2" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModule\AspNetCore\AspNetCore.vcxproj"  Platform="x64" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModule\AspNetCore\AspNetCore.vcxproj" Platform="x86" />
</ItemGroup>

<ItemGroup Condition=" '@(Reference->AnyHaveMetadataValue('Identity', 'AspNetCoreModuleV2'))' == 'true' ">
        <Reference Remove="AspNetCoreModuleV2" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj"  Platform="x64" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj" HandlerPath="2.0.0"  Platform="x64" />

        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\AspNetCore\AspNetCore.vcxproj" Platform="x86" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\OutOfProcessRequestHandler\OutOfProcessRequestHandler.vcxproj" HandlerPath="2.0.0" Platform="x86" />
</ItemGroup>

<ItemGroup Condition=" '@(Reference->AnyHaveMetadataValue('Identity', 'AspNetCoreModuleInProcess'))' == 'true' ">
        <Reference Remove="AspNetCoreModuleInProcess" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj"  Platform="x64" />
        <NativeProjectReference Include="$(MSBuildThisFileDirectory)AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj"  Platform="x86" />
</ItemGroup>
```

File: src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj [hunk 4]
```diff
@@ -5,13 +5,11 @@
     <TargetFramework>netstandard1.0</TargetFramework>
     <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <RestorePackagesPath>$(RepositoryRoot).deps\ANCM</RestorePackagesPath>
```
Comment (natemcmaster) [inline src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj]: I think we can remove this entire project. I was only used as a dummy project to restore packages, and this project no longer has any PackageReferences.

File: src/Installers/Windows/AspNetCoreModule-Setup/Directory.Build.props [hunk 5]
```diff
@@ -31,15 +29,34 @@
...
+    <_ServerIISBasePath>$(RepositoryRoot)\src\Servers\IIS\</_ServerIISBasePath>
     <DefineConstants>BLDVERMAJOR=$(BLDVERMAJOR);BLDVERMINOR=$(BLDVERMINOR);BLDNUMMAJOR=$(BLDNUMMAJOR);BLDNUMMINOR=$(BLDNUMMINOR);$(DefineConstants)</DefineConstants>
     <DefineConstants>ANCMMsiVersion=$(ANCMMsiVersion);ANCMOutOfProcessNugetPackageHandlerVersion=$(ANCMOutOfProcessNugetPackageHandlerVersion);$(DefineConstants)</DefineConstants>
-    <DefineConstants>PreBuiltANCMRoot=$(PreBuiltANCMRoot);PreBuiltANCMV2Root=$(PreBuiltANCMV2Root);$(DefineConstants)</DefineConstants>
-    <DefineConstants>PreBuiltANCMSchema=$(PreBuiltANCMSchema);PreBuiltANCMV2Schema=$(PreBuiltANCMV2Schema);$(DefineConstants)</DefineConstants>
+    <DefineConstants>
+      AspNetCoreSchemaPath=$(_ServerIISBasePath)AspNetCoreModuleV2\AspNetCore\aspnetcore_schema_v2.xml;
+      AspNetCoreMofPath=$(_ServerIISBasePath)AspNetCoreModuleV2\AspNetCore\ancm.mof;
+
+      AspNetCoreV1TargetPath=$(_ServerIISBasePath)AspNetCoreModuleV1\AspNetCore\bin\$(Configuration)\Win32\aspnetcore.dll;
```
Comment (natemcmaster) [inline src/Installers/Windows/AspNetCoreModule-Setup/Directory.Build.props]: Have you considered using ProjectReference instead? The WixToolset will automatically take care of figuring out the path to these files, and this is usually less fragile than hard coding expectations about the location and structure of another project's bin/ directory.

Docs: http://wixtoolset.org/documentation/manual/v3/votive/votive_project_references.html

File: src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj [hunk 6]
```diff
@@ -26,7 +26,7 @@
 
   <ItemGroup Condition="'$(OS)' == 'Windows_NT' AND '$(VCTargetsPath)' != ''">
     <NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="x64" />
-    <NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="x86" />
+    <NativeProjectReference Include="$(MSBuildThisFileDirectory)..\..\AspNetCoreModuleV2\InProcessRequestHandler\InProcessRequestHandler.vcxproj" Platform="Win32" />
```
Comment (natemcmaster) [inline src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj]: Why the change? We've been using x86 everywhere else. If we need to make this change, make sure _everything_ is updated to set Platform=Win32, not x86

File: src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj [hunk 7]
```diff
@@ -24,8 +24,6 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Reference Include="AspNetCoreModule" Condition="'$(OS)' == 'Windows_NT'" />
```
Comment (natemcmaster) [inline src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: Doesn't this break tests? I thought we needed local copies of ancm bits here.

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (natemcmaster) [reply src/Servers/IIS/ResolveIisReferences.targets]: Why are we removing this? Was isn't working correctly?

File: src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj (hunk 6 above)
Comment (pakrym) [reply src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj]: I'll make sure everything is Win32 in native project references.

File: src/Installers/Windows/AspNetCoreModule-Setup/Directory.Build.props (hunk 5 above)
Comment (pakrym) [reply src/Installers/Windows/AspNetCoreModule-Setup/Directory.Build.props]: I, honestly, want to limit the scope of this change. It's getting too big anyway.

File: src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj (hunk 4 above)
Comment (pakrym) [reply src/Installers/Windows/AspNetCoreModule-Setup/ANCMPackageResolver/ANCMPackageResolver.csproj]: Yep

File: src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj (hunk 6 above)
Comment (pakrym) [reply src/Servers/IIS/IIS/src/Microsoft.AspNetCore.Server.IIS.csproj]: Native projects and IIS solution use Win32 platform, there was a fallback in default C++ targets that allowed things to function but it wasn't done correctly.

File: src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj (hunk 7 above)
Comment (pakrym) [reply src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: IIS.IntegrationTesting package would bring required native dependencies.

File: src/Installers/Windows/AspNetCoreModule-Setup/Directory.Build.props (hunk 5 above)
Comment (natemcmaster) [reply src/Installers/Windows/AspNetCoreModule-Setup/Directory.Build.props]: I'm not afraid of bigger PRs, especially if the code is better. I think ProjectReference is safer and is worth doing, and am opposed to not doing it just because this PR has been open for  17 hours.

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (pakrym) [reply src/Servers/IIS/ResolveIisReferences.targets]: Because people don't understand where to use it (almost nowhere) and I don't think there is still need for this magic.

File: src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj (hunk 7 above)
Comment (natemcmaster) [reply src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: How? Did you add something to make the native assets get copied transitively? Or is the SDK automatically handling transitive project refs to native projects?

Comment (pakrym) [reply src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: It was always working because of https://github.com/aspnet/AspNetCore/blob/master/src/Servers/IIS/IntegrationTesting.IIS/src/Microsoft.AspNetCore.Server.IntegrationTesting.IIS.csproj#L39-L43

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (natemcmaster) [reply src/Servers/IIS/ResolveIisReferences.targets]: Which people? The goal of this code was to simplify what it takes to reference native projects. I'm not sure how NativeProjectReference is any less magical.

Comment (natemcmaster) [reply src/Servers/IIS/ResolveIisReferences.targets]: Talked in person. Let's make NativeProjectReference part of eng/targets/ResolveReferences.targets and document it in docs/ReferenceResolution.md

File: src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj (hunk 7 above)
Comment (natemcmaster) [reply src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: That target ensures the native libraries are copied into the build output of Microsoft.AspNetCore.Server.IntegrationTesting.IIS, but how does this make it into the build output of projects which reference Microsoft.AspNetCore.Server.IntegrationTesting.IIS? I don't see anything that would copy the native libraries into the build output of IIS.ForwardsCompatibility.FunctionalTests.

Or,is this just another case of a project that didn't need to have ancm binaries in build output in the first place?

Comment (pakrym) [reply src/Servers/IIS/IIS/test/IIS.ForwardsCompatibility.FunctionalTests/IIS.ForwardsCompatibility.FunctionalTests.csproj]: As far as I remember `<None` item group was copied to output directories of projects referencing `Microsoft.AspNetCore.Server.IntegrationTesting.IIS`. We had this working long before `<NativeProjectReference` came along

File: src/Servers/IIS/ResolveIisReferences.targets (hunk 3 above)
Comment (pakrym) [reply src/Servers/IIS/ResolveIisReferences.targets]: I ended up not needing to use `NativeProjectReference` in WIX projects

File: src/Installers/Windows/AspNetCoreModule-Setup/build/settings/debug.props [hunk 8]
```diff
@@ -14,15 +14,13 @@
     <ClCompile>
       <StringPooling>true</StringPooling>
       <Optimization>Disabled</Optimization>
-      <MinimalRebuild>true</MinimalRebuild>
```
Comment (pakrym) [inline src/Installers/Windows/AspNetCoreModule-Setup/build/settings/debug.props]: Cleaned up build warnings.

**Comment by pakrym** [comment]
:up::date:

File: src/Servers/IIS/build/assets.props [hunk 9]
```diff
@@ -9,263 +9,261 @@
...
-          VCPlatform="x64"
-          Platform="x64"
-          ProjectName="AspNetCore.vcxproj"
-          NativeAsset="aspnetcore"
-          PropetyName="AspNetCoreModuleV1Shim"
-          BaseOutputPath="AspNetCoreModuleV1"
-          ProjectDirectory="$(MSBuildThisFileDirectory)..\AspNetCoreModuleV1\AspNetCore"
-          DllLocation="$(MSBuildThisFileDirectory)..\AspNetCoreModuleV1\AspNetCore\bin\$(Configuration)\x64\aspnetcore.dll"
-          Include="$(MSBuildThisFileDirectory)..\AspNetCoreModuleV1\AspNetCore\AspNetCore.vcxproj"
+          VCPlatform="x64"
```
Comment (jkotalik) [inline src/Servers/IIS/build/assets.props]: Any reason why extra spaces were appended to each line?

Comment (pakrym) [reply src/Servers/IIS/build/assets.props]: A side effect of how the script works. I wasn't able to find a good way to remove them.

**Comment by pakrym** [comment]
:up::date:. This is ready to go in.

**Review by natemcmaster — APPROVED** [review]
